### PR TITLE
Fix end location of compound statements

### PR DESCRIPTION
**Pull request by harupy**
For example, `FunctionDef` currently ends at where `DEDENT` occurs:

```python
def f1():
    pass



def f2():
^ here
    pass
```

but it should end at the final statement in the function body:

```python
def f1():
    pass
       ^ should end here



def f2():
    pass
```

### CODE REVIEW

File: compiler/parser/python.lalrpop [hunk 1]
```diff
@@ -646,15 +647,16 @@ KwargParameter<ArgType>: Option<Box<ast::Arg>> = {
 };
 
 ClassDef: ast::Stmt = {
-    <decorator_list:Decorator*> <location:@L> "class" <name:Identifier> <a:("(" ArgumentList ")")?> ":" <body:Suite> <end_location:@R> => {
+    <decorator_list:Decorator*> <location:@L> "class" <name:Identifier> <a:("(" ArgumentList ")")?> ":" <body:Suite> => {
```
Comment (harupy) [inline compiler/parser/python.lalrpop:650]: Just modified `ClassDef` and `FuncDef` as a test.

Comment (harupy) [reply compiler/parser/python.lalrpop:650]: @DimitrisJim Thanks for the comment. Yes, I need to apply the same change to other compound statements. I'm working on that now.

Comment (DimitrisJim) [reply compiler/parser/python.lalrpop:650]: ah yes, just saw it's still a draft! :smile:

Comment (harupy) [reply compiler/parser/python.lalrpop:681]: @DimitrisJim Updated other compound statements.

File: compiler/parser/python.lalrpop [hunk 2]
```diff
@@ -416,10 +434,19 @@ TryStatement: ast::Stmt = {
...
         let finalbody = finally.map(|s| s.2).unwrap_or_default();
+        let end_location = if let Some(last) = finalbody.last() {
+            last.end_location
+        } else {
+            if let Some(last) = orelse.last() {
+                last.end_location
+            } else {
+                handlers.last().unwrap().end_location
+            }
+        };
```
Comment (harupy) [inline compiler/parser/python.lalrpop:445]: Asked ChatGPT if we can simplify this code:

image

File: compiler/parser/python.lalrpop [hunk 3]
```diff
@@ -375,19 +379,23 @@ IfStatement: ast::Stmt = {
...
     },
 };
 
 WhileStatement: ast::Stmt = {
-    <location:@L> "while" <test:NamedExpressionTest> ":" <body:Suite> <s2:("else" ":" Suite)?> <end_location:@R> => {
+    <location:@L> "while" <test:NamedExpressionTest> ":" <body:Suite> <s2:("else" ":" Suite)?> => {
+        let end_location = s2
+            .as_ref()
+            .map(|s| s.2.last().unwrap().end_location)
+            .unwrap_or_else(|| body.last().unwrap().end_location);
```
Comment (DimitrisJim) [inline compiler/parser/python.lalrpop:393]: there's the `orelse` declared in the next line (line `394`), move that to the front and use the same logic as the other places, i.e:

```suggestion
        let orelse = s2.map(|s| s.2).unwrap_or_default();
        let end_location = orelse
            .last()
            .or_else(body.last())
            .unwrap()
            .end_location;
```

File: compiler/parser/python.lalrpop [hunk 4]
```diff
@@ -398,7 +406,12 @@ WhileStatement: ast::Stmt = {
 };
 
 ForStatement: ast::Stmt = {
-    <location:@L> <is_async:"async"?> "for" <target:ExpressionList> "in" <iter:TestList> ":" <body:Suite> <s2:("else" ":" Suite)?> <end_location:@R> => {
+    <location:@L> <is_async:"async"?> "for" <target:ExpressionList> "in" <iter:TestList> ":" <body:Suite> <s2:("else" ":" Suite)?> => {
+        let end_location = s2
```
Comment (DimitrisJim) [inline compiler/parser/python.lalrpop:410]: similar case here as in the `While` case.

**Review by DimitrisJim — COMMENTED** [review]
thanks! I've left a couple of comments on this

**Comment by harupy** [comment]
@DimitrisJim Thanks for the comments! Applied your suggestions.

**Review by DimitrisJim — APPROVED** [review]
looks good to me, thanks!

**Comment by harupy** [comment]
cc @charliermarsh

**Pull request merged**